### PR TITLE
Fix/Improve `processWindowErrorEvent`

### PR DESCRIPTION
**Pull request by silverwind**
- `e.error` can be undefined in some cases which would raise an error inside this error handler, fixed that.
- The displayed message mentions looking into the console, but in my case of error from `ResizeObserver` there was nothing there, so add this logging. I think this logging was once there but got lost during refactoring.

### CODE REVIEW

**Comment by silverwind** [comment]
Maybe browser does the logging, but only if there is `e.error`, need to test.

**Review by yardenshoham — APPROVED** [review]

**Comment by silverwind** [comment]
Putting draft until I have confirmed above behaviour. Definitely should backport this, otherwise some "errors" will cause errors in inside this error handler.

**Review by delvh — APPROVED** [review]

**Comment by silverwind** [comment]
Rewrote the function, now it truly handles all cases and the comments accurately describe what it does.

Flashes:
Screenshot 2024-02-26 at 20 14 57

Console.
Screenshot 2024-02-26 at 20 15 04

File: web_src/js/bootstrap.js [hunk 1]
```diff
@@ -28,27 +28,30 @@ export function showGlobalErrorMessage(msg) {
...
+  if (!err && message) console.error(new Error(message));
 
-  let message;
-  if (e.type === 'unhandledrejection') {
-    message = `JavaScript promise rejection: ${err.message}.`;
-  } else {
-    message = `JavaScript error: ${e.message} (${e.filename} @ ${e.lineno}:${e.colno}).`;
-  }
+  // If the error stack trace does not include the base URL of our scripts, it is likely coming from
+  // a browser extension or inline script. Do not show these in production builds.
```
Comment (silverwind) [inline web_src/js/bootstrap.js]:
```suggestion
  // a browser extension or inline script. Do not show such errors in production builds.
```

File: web_src/js/bootstrap.js [hunk 2]
```diff
@@ -28,27 +28,32 @@ export function showGlobalErrorMessage(msg) {
...
-    return; // ignore such nonsense error event
-  }
+  // If the error stack trace does not include the base URL of our script assets, it likely came
+  // from a browser extension or inline script. Do not show such errors in production.
+  if (!err?.stack?.includes(assetBaseUrl) && runModeIsProd) return;
 
-  showGlobalErrorMessage(`${message} Open browser console to see more details.`);
+  const renderedType = type === 'unhandledrejection' ? 'promise rejection' : type;
+  let msg = err?.message ?? message;
+  if (lineno) msg += `(${filename} @ ${lineno}:${colno})`;
```
Comment (silverwind) [inline web_src/js/bootstrap.js]:
```suggestion
  if (lineno) msg += ` (${filename} @ ${lineno}:${colno})`;
```

File: web_src/js/bootstrap.js [hunk 3]
```diff
@@ -28,27 +28,32 @@ export function showGlobalErrorMessage(msg) {
...
-
-  let message;
-  if (e.type === 'unhandledrejection') {
-    message = `JavaScript promise rejection: ${err.message}.`;
-  } else {
-    message = `JavaScript error: ${e.message} (${e.filename} @ ${e.lineno}:${e.colno}).`;
+  // This handler not only receives errors but also error events, detectable by `err` being null
+  // or undefined. Error events do not log to the browser console by default but as they might
+  // still be relevant, we show them during development. In production they will only log to the
+  // console. References and examples:
```
Comment (silverwind) [inline web_src/js/bootstrap.js]:
```suggestion
  // still be relevant, we show them during development. References and examples:
```

File: web_src/js/bootstrap.js [hunk 4]
```diff
@@ -28,27 +29,31 @@ export function showGlobalErrorMessage(msg) {
...
 
-  // error is likely from browser extension or inline script. Do not show these in production builds.
-  if (!err.stack?.includes(assetBaseUrl) && window.config?.runModeIsProd) return;
-
-  let message;
-  if (e.type === 'unhandledrejection') {
-    message = `JavaScript promise rejection: ${err.message}.`;
-  } else {
-    message = `JavaScript error: ${e.message} (${e.filename} @ ${e.lineno}:${e.colno}).`;
+  // This handler not only receives `Errors`` but also error events, detectable by `error` being
```
Comment (silverwind) [inline web_src/js/bootstrap.js]:
```suggestion
  // This handler not only receives `Error`s but also error events, detectable by `error` being
```

**Review by wxiaoguang — COMMENTED** [review]
I don't like this change. Because "e" could be anything by design.

But I won't block it.

**Comment by silverwind** [comment]
> I don't like this change. Because "e" could be anything by design.
> 
> But I won't block it.

I know you don't like such destructuring, but [ErrorEvent](https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent) and its properties are well defined, so I see no issue.

**Comment by silverwind** [comment]
The handler also receives [PromiseRejectionEvent](https://developer.mozilla.org/en-US/docs/Web/API/PromiseRejectionEvent) and I see its `reason` property is not guaranteed to be an `Error`, so it could be improved to handle that case.

**Comment by wxiaoguang** [comment]
> I know you don't like such destructuring

Not really. The real problem is `Because "e" could be anything by design.`. Before the code was quite straightforward: if this error, do this, if that error, do that. But the new code just mixed everything together. I don't like making readers "guess" what the type/value there should be.

**Comment by wxiaoguang** [comment]
ps: the last time I said "no" to such syntax, the reason is the same: it make code unclear (https://github.com/go-gitea/gitea/pull/25253#discussion_r1229359091), and introduced new bugs https://github.com/go-gitea/gitea/pull/25253#discussion_r1230361844

**Comment by silverwind** [comment]
I can restore the previous `type` check, no problem.

**Comment by silverwind** [comment]
I decided against `type` check as there is no overlap in properties between ErrorEvent and PromiseRejectionEvent, so it would unnecessarily bloat the code. I fixed the jsDoc, added a typecheck for `Error` before accessing `stack`, and tweaked the comment again.

Everything works as expected:

Screenshot 2024-02-27 at 21 18 52

**Comment by silverwind** [comment]
> Before the old code (before https://github.com/go-gitea/gitea/pull/29303) was quite straightforward: if this error, do this, if that error, do that. But the new code just mixed everything together. I don't like making readers "guess" what the type/value there should be.

It's pretty straightforward already. Could be made even more so by splitting into two handler functions hat converge into a logging function, but I think this be in overengineering territory. I don't see the current function as hard to read just because of this `error ?? reason`, both serve the same purpose.

**Comment by silverwind** [comment]
@delvh @yardenshoham maybe you want to give this another look as it has changed a bit since your reviews. If there no further objections, I will merge this soon.

File: web_src/js/bootstrap.js [hunk 5]
```diff
@@ -26,29 +27,42 @@ export function showGlobalErrorMessage(msg) {
 }
 
 /**
- * @param {ErrorEvent} e
+ * @param {ErrorEvent|PromiseRejectionEvent} event- Event
```
Comment (silverwind) [inline web_src/js/bootstrap.js]:
```suggestion
 * @param {ErrorEvent|PromiseRejectionEvent} event - Event
```

**Comment by yardenshoham** [comment]
Please recommend ways to test

**Comment by silverwind** [comment]
Can be tested by installing Violentmonkey with such userscripts:

```js
// ==UserScript==
// @name        Gitea error trigger 1
// @match       http://localhost:3000/*
// ==/UserScript==
setTimeout(() => {
  throw new Error("sync error from userscript");
}, 10000)

const divElem = document.querySelector(".repo-header");
const resizeObserver = new ResizeObserver((entries) => {
  for (const entry of entries) {
    entry.target.style.width = entry.contentBoxSize[0].inlineSize + 10 + "px";
  }
});
resizeObserver.observe(divElem);
```
```js
// ==UserScript==
// @name        Gitea error trigger 2
// @match       http://localhost:3000/*
// ==/UserScript==
(async () => {
  throw new Error("async error from userscript");
})()
```

**Comment by yardenshoham** [comment]
I can confirm I see the errors
image
image

**Comment by silverwind** [comment]
If you go on a repo page, you should also see the ResizeObserver error :)

**Pull request merged**